File: original/js/drawEnv.js
export class DrawEnv {
  constructor() {
    this.nowSin = 0;
    this.nowCos = 1;
    this.width = 320;
    this.height = 200;
    this.imageData = null;
    this.pixelBuffer = null;
  }

  initBuffer() {
    this.imageData = new ImageData(this.width, this.height);
    this.pixelBuffer = this.imageData.data;
  }

  clearBuffer(r, g, b) {
    const buf = this.pixelBuffer;
    const len = buf.length;
    for (let i = 0; i < len; i += 4) {
      buf[i]     = r;
      buf[i + 1] = g;
      buf[i + 2] = b;
      buf[i + 3] = 255;
    }
  }

  flush(ctx) {
    ctx.putImageData(this.imageData, 0, 0);
  }

  _project(p) {
    const scale = 120 / (1 + 0.6 * p.z);
    const rotX = this.nowCos * p.x + this.nowSin * (p.y - 2.0);
    const rotY = -this.nowSin * p.x + this.nowCos * (p.y - 2.0) + 2.0;
    return {
      x: ((rotX * scale) | 0) + 160,
      y: ((rotY * scale) | 0) + 100
    };
  }

  // Pixel-exact scanline fill — matches Java's fillPolygon behavior
  _fillPolygon(screenPts, r, g, b) {
    const buf = this.pixelBuffer;
    const W = this.width;
    const H = this.height;
    const n = screenPts.length;

    let yMin = H, yMax = -1;
    for (let i = 0; i < n; i++) {
      const y = screenPts[i].y;
      if (y < yMin) yMin = y;
      if (y > yMax) yMax = y;
    }
    yMin = Math.max(0, yMin | 0);
    yMax = Math.min(H - 1, yMax | 0);

    for (let y = yMin; y <= yMax; y++) {
      let xLeft = W, xRight = -1;
      for (let i = 0; i < n; i++) {
        const v1 = screenPts[i];
        const v2 = screenPts[(i + 1) % n];
        if (v1.y === v2.y) continue; // skip horizontal edges
        const yLo = v1.y < v2.y ? v1.y : v2.y;
        const yHi = v1.y < v2.y ? v2.y : v1.y;
        if (y >= yLo && y <= yHi) {
          const x = (v1.x + (y - v1.y) * (v2.x - v1.x) / (v2.y - v1.y)) | 0;
          if (x < xLeft) xLeft = x;
          if (x > xRight) xRight = x;
        }
      }
      xLeft = Math.max(0, xLeft);
      xRight = Math.min(W - 1, xRight);
      for (let x = xLeft; x <= xRight; x++) {
        const idx = (y * W + x) << 2;
        buf[idx]     = r;
        buf[idx + 1] = g;
        buf[idx + 2] = b;
        buf[idx + 3] = 255;
      }
    }
  }

  drawFace(face) {
    const pts = face.points;
    const dx1 = pts[1].x - pts[0].x;
    const dy1 = pts[1].y - pts[0].y;
    const dx2 = pts[2].x - pts[0].x;
    const dy2 = pts[2].y - pts[0].y;
    let brightness = Math.fround(Math.abs(dx1 * dy2 - dy1 * dx2) / face.maxZ);
    if (brightness > 1) brightness = 1;

    const r = Math.round(face.red * brightness * 255);
    const g = Math.round(face.green * brightness * 255);
    const b = Math.round(face.blue * brightness * 255);

    const n = face.numPoints;
    const screenPts = new Array(n);
    for (let i = 0; i < n; i++) {
      screenPts[i] = this._project(pts[i]);
    }
    this._fillPolygon(screenPts, r, g, b);
  }

  drawPolygon(color, points) {
    const n = points.length;
    const screenPts = new Array(n);
    for (let i = 0; i < n; i++) {
      screenPts[i] = this._project(points[i]);
    }
    this._fillPolygon(screenPts, color.r, color.g, color.b);
  }
}

File: decomp/DrawEnv.java
import java.awt.Color;
import java.awt.Graphics;

public class DrawEnv {
   static final double T = 0.6D;
   private int[] polyX = new int[8];
   private int[] polyY = new int[8];
   private DPoint3[] dp3 = new DPoint3[3];
   double nowSin;
   double nowCos;
   int width;
   int height;

   synchronized void drawPolygon(Graphics var1, Face var2) {
      int var3 = var2.numPoints;
      DPoint3[] var4 = var2.points;
      double var6 = var4[1].x - var4[0].x;
      double var8 = var4[1].y - var4[0].y;
      DPoint3 var10000 = var4[1];
      var10000 = var4[0];
      double var10 = var4[2].x - var4[0].x;
      double var12 = var4[2].y - var4[0].y;
      var10000 = var4[2];
      var10000 = var4[0];
      float var14 = (float)(Math.abs(var6 * var12 - var8 * var10) / var2.maxZ);
      var1.setColor(new Color(var2.red * var14, var2.green * var14, var2.blue * var14));
      double var15 = (double)this.width / 320.0D;
      double var17 = (double)this.height / 200.0D;

      for(int var5 = 0; var5 < var3; ++var5) {
         DPoint3 var19 = var4[var5];
         double var20 = 120.0D / (1.0D + 0.6D * var19.z);
         double var22 = this.nowCos * var19.x + this.nowSin * (var19.y - 2.0D);
         double var24 = -this.nowSin * var19.x + this.nowCos * (var19.y - 2.0D) + 2.0D;
         this.polyX[var5] = (int)(var22 * var15 * var20) + this.width / 2;
         this.polyY[var5] = (int)(var24 * var17 * var20) + this.height / 2;
      }

      var1.fillPolygon(this.polyX, this.polyY, var3);
   }

   synchronized void drawPolygon(Graphics var1, DPoint3[] var2) {
      int var3 = var2.length;
      double var5 = (double)this.width / 320.0D;
      double var7 = (double)this.height / 200.0D;

      for(int var4 = 0; var4 < var3; ++var4) {
         DPoint3 var9 = var2[var4];
         double var10 = 120.0D / (1.0D + 0.6D * var9.z);
         double var12 = this.nowCos * var9.x + this.nowSin * (var9.y - 2.0D);
         double var14 = -this.nowSin * var9.x + this.nowCos * (var9.y - 2.0D) + 2.0D;
         this.polyX[var4] = (int)(var12 * var5 * var10) + this.width / 2;
         this.polyY[var4] = (int)(var14 * var7 * var10) + this.height / 2;
      }

      var1.fillPolygon(this.polyX, this.polyY, var3);
   }
}
